src/libraries/System.Private.CoreLib/src/System/Environment.Win32.cs

```diff
@@ -364,18 +372,17 @@ private static string GetFolderPathCore(SpecialFolder folder, SpecialFolderOptio
                     break;
                 case SpecialFolder.Windows:
                     folderGuid = Interop.Shell32.KnownFolders.Windows;
+                    fallbackEnv = "windir";
                     break;
             }
 
             Guid folderId = new Guid(folderGuid);
-
             int hr = Interop.Shell32.SHGetKnownFolderPath(folderId, (uint)option, IntPtr.Zero, out string path);
```

Comment:
This function is part of the shell experience, so desktop only. So I'm not surprised that it fails on systems without shell. This will probably also fix usages in system services, etc. And probably everything else where no user registry hives are loaded.